Auto-retry sending the message when posting fails.

zoomix opened the pull request:
#### Summary
Introduced an incremental backoff for sending messages, if a message fails. This saves the user clicking the "retry" when posting the message fails. Nothing visible to the user.

There's a __DEV__ statement to show a toast if the fault is triggered in emulator mode. 

DRYed up the server error handling.

#### Ticket Link

https://github.com/mattermost/mattermost-mobile/issues/8964 Once every few messages, mattermost-mobile on android will fail to send a message. A retry immediately works. See video.

#### Device Information
This PR was tested on: 
Android 15 emulator
Motorola Razr 50 Ultra


#### Release Note

```release-note
Auto-retries sending the message if failure is detected
```

mattermost-build commented:
Hello @zoomix,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

zoomix commented:
Hi @rahimrahman ,

No problem, thanks for reviewing it. I've been running my own build since the PR so I'd scratched my itch. =)

I agree that retry would probably be best handled in the client, but if I were to guess why it was there it would be because of the potential that the error was caused by server state. The bit about the error handling I refactored. This:
https://github.com/mattermost/mattermost-mobile/blob/main/app/actions/remote/post.ts#L157-L163

.. or the method I extracted it into ( https://github.com/zoomix/mattermost-mobile/blob/04b8f207b2ebe61325e3646d26e932f5d94da631/app/actions/remote/post.ts#L65-L73 )

But maybe this itself is not the way you wanna handle this kind of error? Or the problem isn't relevant? I'm unsure.

enahum commented:
The reason there is no automatic retry on create post is cause there was a problem with idempotency on the server, specially in an HA environment. Not sure if that was ever addressed, of it was, it could simply be enabled and adjusted

rahimrahman commented:
> Hi @rahimrahman ,
> 
> No problem, thanks for reviewing it. I've been running my own build since the PR so I'd scratched my itch. =)
> 
> I agree that retry would probably be best handled in the client, but if I were to guess why it was there it would be because of the potential that the error was caused by server state. The bit about the error handling I refactored. This: https://github.com/mattermost/mattermost-mobile/blob/main/app/actions/remote/post.ts#L157-L163
> 
> .. or the method I extracted it into ( https://github.com/zoomix/mattermost-mobile/blob/04b8f207b2ebe61325e3646d26e932f5d94da631/app/actions/remote/post.ts#L65-L73 )
> 
> But maybe this itself is not the way you wanna handle this kind of error? Or the problem isn't relevant? I'm unsure.

Ah, I see.  That make sense.

I think we're currently doing EXPONENTIAL_RETRY, which means it should be [1000, 2000, 4000], but I'm not too worried about it.

The only thing I would love see is unit test on the changes you introduced, if that's something you can include.  The easiest way is to extract the functions out and test those individual functions: `isErrorThatWarrantsNotPostingTheMessage` and `newFunctionThatDoesFetchingWithRetry`.

zoomix commented:
Something like this?

I was debating with myself if I should chuck the `createPostWithRetries` and `isErrorThatWarrantsNotPostingTheMessage` in with the `create, update & delete posts` bundle, but ultimately decided against it since the methods themselves shouldn't be used by any of their peers. 

There were already tests for the `isErrorThatWarrantsNotPostingTheMessage`, but they were implicit in so far as they were specced for the greater `createPost` method. 

So ultimately I decided to wrap them individually in their own `describe` s. I had a (non-exhaustive) look at the other tests and could not find a suitable example of how these kinds of methods are usually tested.

enahum commented:
I've triggered the tests.

Now, I wonder how does this PR prevents postial duplicates of the same message in an HA environment?

larkox commented:
@zoomix and sorry if this has been discussed already but... why adding the whole logic instead of removing from `app/client/rest/posts.ts` line 43 the `noRetry` option?

rahimrahman commented:
> The reason there is no automatic retry on create post is cause there was a problem with idempotency on the server, specially in an HA environment. Not sure if that was ever addressed, if it was, the auto retries could simply be enabled and adjusted

@larkox to answer your question on the noRetry.

larkox commented:
@zoomix I am closing this since we are going with https://github.com/mattermost/react-native-network-client/pull/148 .

If I misunderstood something, feel free to reopen the PR 😄
